### PR TITLE
Enhances parsing of strings into integer/longs

### PR DESCRIPTION
**Pull request by idlira**
By considering values with only zeros as after the decimal point (or a maximal epsilon of 0.000001d).

See added test cases for examples.

> [OX-9546](https://scireum.myjetbrains.com/youtrack/issue/OX-9546)
> [OX-9560](https://scireum.myjetbrains.com/youtrack/issue/OX-9560)

### CODE REVIEW

**Review by sabieber — APPROVED** [review]

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 1]
```diff
@@ -1169,6 +1190,38 @@ private static <V> V parseBasicTypesFromUserString(Class<V> clazz, String value,
         return parseDatesFromUserString(clazz, value, language);
     }
 
+    private static Integer parseIntegerFromUser(String value, String language) {
+        try {
+            Double valueAsDouble = parseDecimalNumberFromUser(value, language);
```
Comment (jakobvogel) [inline src/main/java/sirius/kernel/nls/NLS.java:1195]: Probably I miss something, but isn't this exactly the code above, just nested in another catch clause? Why not invoke this?

And if it is better to copy the code, why not fire the desired exception class directly?

**Review by jakobvogel — COMMENTED** [review]
I am puzzled... Please help me to understand the change 🤗

File: src/main/java/sirius/kernel/nls/NLS.java (hunk 1 above)
Comment (idlira) [reply src/main/java/sirius/kernel/nls/NLS.java:1195]: There are basically 4 variants, slightly close but not the same.
- Integer x Long - different internal types
- FromUserString x FromMachineString - for the first we reuse the parseDoubleFromUserString, which throws a different exception as the Java's valueOf used in FromMachineString

**Review by jakobvogel — APPROVED** [review]
Ah, ok, now I see! 👍

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 2]
```diff
@@ -1025,14 +1026,14 @@ public static <V> V parseMachineString(Class<V> clazz, String value) {
...
             } catch (NumberFormatException e) {
                 throw new IllegalArgumentException(fmtr("NLS.errInvalidNumber").set("value", value).format(), e);
             }
         }
         if (Long.class.equals(clazz) || long.class.equals(clazz)) {
             try {
-                return (V) Long.valueOf(value);
+                return (V) parseLongFromMachineString(value);
             } catch (NumberFormatException e) {
                 throw new IllegalArgumentException(fmtr("NLS.errInvalidNumber").set("value", value).format(), e);
```
Comment (andyHa) [inline src/main/java/sirius/kernel/nls/NLS.java:1038]: same as above but .toLongValue

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 3]
```diff
@@ -1025,14 +1026,14 @@ public static <V> V parseMachineString(Class<V> clazz, String value) {
     private static <V> V parseBasicTypesFromMachineString(Class<V> clazz, String value) {
         if (Integer.class.equals(clazz) || int.class.equals(clazz)) {
             try {
-                return (V) Integer.valueOf(value);
+                return (V) parseIntegerFromMachineString(value);
             } catch (NumberFormatException e) {
                 throw new IllegalArgumentException(fmtr("NLS.errInvalidNumber").set("value", value).format(), e);
```
[the remark this answers is not in the document]
Comment (andyHa) [reply src/main/java/sirius/kernel/nls/NLS.java:1031]: refactor this "fail is factional part into a method" and re-use 4 times

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 4]
```diff
@@ -1136,14 +1157,14 @@ public static <V> V parseUserString(Class<V> clazz, String value, String languag
     private static <V> V parseBasicTypesFromUserString(Class<V> clazz, String value, String language) {
         if (Integer.class.equals(clazz) || int.class.equals(clazz)) {
             try {
-                return (V) Integer.valueOf(value);
+                return (V) parseIntegerFromUser(value, language);
             } catch (NumberFormatException e) {
                 throw new IllegalArgumentException(fmtr("NLS.errInvalidNumber").set("value", value).format(), e);
```
[the remark this answers is not in the document]
Comment (andyHa) [reply src/main/java/sirius/kernel/nls/NLS.java:1162]: this would also add support for machine formats (see below)

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 5]
```diff
@@ -1136,14 +1157,14 @@ public static <V> V parseUserString(Class<V> clazz, String value, String languag
...
             }
         }
         if (Long.class.equals(clazz) || long.class.equals(clazz)) {
             try {
-                return (V) Long.valueOf(value);
+                return (V) parseLongFromUser(value, language);
             } catch (NumberFormatException e) {
                 throw new IllegalArgumentException(fmtr("NLS.errInvalidNumber").set("value", value).format(), e);
```
Comment (andyHa) [inline src/main/java/sirius/kernel/nls/NLS.java:1169]: s.a.

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 6]
```diff
@@ -1169,17 +1190,51 @@ private static <V> V parseBasicTypesFromUserString(Class<V> clazz, String value,
...
-        } catch (ParseException exception) {
-            Exceptions.ignore(exception);
-            return Double.valueOf(value);
+            try {
+                return getDecimalFormat(language).parse(value).doubleValue();
+            } catch (ParseException exception) {
+                Exceptions.ignore(exception);
+                return Double.valueOf(value);
+            }
         } catch (NumberFormatException exception) {
```
Comment (andyHa) [inline src/main/java/sirius/kernel/nls/NLS.java:1238]: i don't understand this change

**Review by sabieber — APPROVED** [review]

File: src/test/java/sirius/kernel/nls/NLSSpec.groovy [hunk 7]
```diff
@@ -148,6 +144,26 @@ class NLSSpec extends BaseSpecification {
...
     }
 
+    def "parseUserString works for integers"() {
+        expect:
+        NLS.parseUserString(Integer.class, input) == output
+
+        where:
+        input     | output
+        "42"      | 42
+        "77,0000" | 77
```
Comment (andyHa) [inline src/test/java/sirius/kernel/nls/NLSSpec.groovy:154]: ensure that german and english 77.000 and 77,000 still yield 77000 and also other edge cases

**Review by andyHa — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/kernel/nls/NLS.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (idlira) [reply src/main/java/sirius/kernel/nls/NLS.java:1031]: Can be done, but the parsing is more complicated. "13,00" will render a BigDecimal with scale 2.

[the remark this answers is not in the document]
Comment (idlira) [reply src/main/java/sirius/kernel/nls/NLS.java:1031]: See new code using the common method `extractNonFactionalPart()`.
I kept the approach of using a double to detect the fractional part. A BigDecimal has no public method to obtain this information.
